File: frontend/src/entities/track/model/track.ts
import { OwnerTrackStateTone, StreamQuality, Track } from '@/shared/api/types';

export function getTrackStatusColor(status: Track['status']): 'default' | 'warning' | 'success' | 'error' {
  if (status === 'approved') {
    return 'success';
  }

  if (status === 'pending' || status === 'processing') {
    return 'warning';
  }

  if (status === 'rejected' || status === 'hidden' || status === 'deleted') {
    return 'error';
  }

  return 'default';
}

export function getTrackStatusLabel(status: Track['status']): string {
  const labels: Record<Track['status'], string> = {
    pending: 'Ждёт аудио',
    processing: 'Обрабатывается',
    approved: 'Опубликован',
    rejected: 'Нужно исправить',
    hidden: 'Скрыт',
    deleted: 'Удалён',
  };

  return labels[status];
}

export function getOwnerTrackState(track: Track): {
  tone: OwnerTrackStateTone;
  title: string;
  description: string;
} {
  const hasMedia = hasPlayableMedia(track);

  if (track.status === 'deleted') {
    return {
      tone: 'error',
      title: 'Трек снят с публикации',
      description: 'Этот трек больше не отображается в каталоге и недоступен слушателям.',
    };
  }

  if (track.status === 'hidden') {
    return {
      tone: 'error',
      title: 'Трек скрыт командой проекта',
      description: 'Запись не видна в публичном каталоге. Заменить аудио или обложку можно после восстановления.',
    };
  }

  if (track.status === 'processing') {
    return {
      tone: 'warning',
      title: 'Аудио обрабатывается',
      description: 'Файл принят. Плеер и публикация станут доступны после завершения обработки.',
    };
  }

  if (track.status === 'rejected') {
    return {
      tone: 'error',
      title: 'Обработка завершилась с ошибкой',
      description: 'Исправьте файл или загрузите новое аудио, чтобы снова отправить трек на обработку.',
    };
  }

  if (track.status === 'approved') {
    return {
      tone: 'success',
      title: 'Трек опубликован автоматически',
      description: 'После успешной обработки запись сразу стала доступна в каталоге и в общем плеере.',
    };
  }

  if (track.status === 'pending' && !hasMedia) {
    return {
      tone: 'info',
      title: 'Ждёт исходный файл',
      description: 'Карточка сохранена. Следующий шаг: загрузить MP3 или WAV, чтобы трек дошёл до публикации.',
    };
  }

  return {
    tone: 'warning',
    title: 'Трек ждёт обновления',
    description: 'Загрузите аудио заново, чтобы трек прошёл актуальную обработку и появился в каталоге.',
  };
}

export function canUploadTrackMedia(track: Track): boolean {
  return track.status === 'pending' || track.status === 'rejected' || track.status === 'approved';
}

export function hasPlayableMedia(track: Track): boolean {
  if (track.status === 'hidden' || track.status === 'deleted') {
    return false;
  }
  return track.status === 'approved' || Boolean(track.original_url || track.mp3_128_url || track.mp3_320_url);
}

export function getPlayableQualityCandidates(track: Track, preferredQuality: StreamQuality): StreamQuality[] {
  const qualities: StreamQuality[] = [];
  const addQuality = (quality: StreamQuality) => {
    if (!qualities.includes(quality)) {
      qualities.push(quality);
    }
  };

  const hasPublicStreamFallback = track.status === 'approved';
  const hasQuality = (quality: StreamQuality) => {
    if (hasPublicStreamFallback) {
      return true;
    }
    if (quality === '320') {
      return Boolean(track.mp3_320_url);
    }
    if (quality === '128') {
      return Boolean(track.mp3_128_url);
    }
    return Boolean(track.original_url);
  };

  if (hasQuality(preferredQuality)) {
    addQuality(preferredQuality);
  }

  const fallbackQualities: StreamQuality[] = ['320', '128', 'original'];
  for (const quality of fallbackQualities) {
    if (hasQuality(quality)) {
      addQuality(quality);
    }
  }

  return qualities;
}

export const emptyTrack: Track = {
  id: 0,
  title: 'R',
  user_id: 0,
  artist_id: 0,
  is_public: true,
  is_downloadable: false,
  license_type: 'all-rights-reserved',
  status: 'approved',
  created_at: '',
  updated_at: '',
  play_count: 0,
  like_count: 0,
  comment_count: 0,
};

File: backend/app/models/track.py
from enum import Enum as PyEnum

from sqlalchemy import ARRAY, JSON, BigInteger, Boolean, Column, DateTime, Enum, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.session import Base


class TrackStatus(str, PyEnum):
    pending = "pending"
    processing = "processing"
    approved = "approved"
    rejected = "rejected"
    hidden = "hidden"
    deleted = "deleted"


class Track(Base):
    __tablename__ = "tracks"
    __table_args__ = (
        Index("ix_tracks_status_created_at", "status", "created_at"),
        Index("ix_tracks_category_status", "category_id", "status"),
        Index("ix_tracks_user_created_at", "user_id", "created_at"),
    )

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text)
    genre = Column(String(100))
    category_id = Column(Integer, ForeignKey("categories.id", ondelete="SET NULL"), index=True)
    duration_seconds = Column(Integer)
    file_size_bytes = Column(BigInteger)
    original_url = Column(Text)
    mp3_128_url = Column(Text)
    mp3_320_url = Column(Text)
    cover_image_url = Column(Text)
    waveform_data_json = Column(JSON)
    metadata_json = Column(JSON)
    status = Column(Enum(TrackStatus, name="track_status"), default=TrackStatus.pending, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    play_count = Column(Integer, default=0, index=True)
    like_count = Column(Integer, default=0, index=True)
    comment_count = Column(Integer, default=0)
    is_public = Column(Boolean, default=True, index=True)
    is_downloadable = Column(Boolean, default=False)
    license_type = Column(String(50), default="all-rights-reserved")
    tags = Column(ARRAY(String))
    bpm = Column(Integer)
    key_signature = Column(String(20))
    rejection_reason = Column(Text)

    user = relationship("User", back_populates="tracks")
    category = relationship("Category", back_populates="tracks")
    interactions = relationship("Interaction", back_populates="track", cascade="all, delete-orphan")
